### PR TITLE
introduce UniversalToJsonFormatterFactory

### PR DESCRIPTION
**Pull request by gadomsky**

### CODE REVIEW

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalMessageFormatter.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import scala.reflect.ClassTag
+
+/**
+ * @param schemaRegistryClient schema registry client
+ */
+private[confluent] class UniversalMessageFormatter(schemaRegistryClient: SchemaRegistryClient) {
+
+  private lazy val avroMessageFormatter = new ConfluentAvroMessageFormatter(schemaRegistryClient)
+
+  def asJson[T: ClassTag](obj: T, parsedSchema: ParsedSchema): Json = parsedSchema match {
```
Comment (mproch) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalMessageFormatter.scala:26]: I think we don't need T parameter here and in ConfluentAvroMessageFormatter.

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalToJsonFormatterFactory.scala [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    case str: String => Json.fromString(str)
+    case key => messageFormatter.asJson[K](key, schemaOpt.get)
+  }
+
+  protected def createValueEncoder(messageFormatter: UniversalMessageFormatter, schema: ParsedSchema): Encoder[V] = (value: V) => messageFormatter.asJson[V](value, schema)
+
+  implicit protected val serializableRecordDecoder: Decoder[SerializableConsumerRecord[Json, Json]] = deriveConfiguredDecoder
+  protected val consumerRecordDecoder: Decoder[UniversalSerializableConsumerRecord[Json, Json]] = deriveConfiguredDecoder
+
+  protected def keyEncoder(schemaOpt: Option[ParsedSchema]): Encoder[K] = createKeyEncoder(messageFormatter, schemaOpt)
```
Comment (mproch) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalToJsonFormatterFactory.scala:117]: why do we need these methods?

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalMessageReader.scala [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  */
+private[confluent] class UniversalMessageReader(schemaRegistryClient: SchemaRegistryClient) {
+
+  private val avroMessageReader = new ConfluentAvroMessageReader(schemaRegistryClient)
+
+  def readJson(jsonObj: Json, parsedSchema: ParsedSchema, subject: String): Array[Byte] = {
+    parsedSchema match {
+      case schema: AvroSchema => avroMessageReader.readJson(jsonObj, schema.rawSchema(), subject)
+      //todo: change JsonSchema
+      case s: ParsedSchema if s.schemaType() == "JSON" =>
```
Comment (mproch) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalMessageReader.scala:21]: could we extract all such cases to one place? 
and have one method like
```
(ParsedSchema, some config?) => {
    asJson(obj, subject): Json
    readJson(json, subject): Array[Byte]
    type: TypingResult
    deserialize(...)...
}
```
so we know what is needed to introduce new schema type?

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalMessageFormatter.scala (hunk 1 above)
Comment (gadomsky) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalMessageFormatter.scala]: removed from both

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalToJsonFormatterFactory.scala (hunk 2 above)
Comment (gadomsky) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalToJsonFormatterFactory.scala]: no :) removed

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/source/UniversalKafkaSourceFactory.scala [hunk 4]
```diff
@@ -104,14 +104,12 @@ class UniversalKafkaSourceFactory[K: ClassTag, V: ClassTag](val schemaRegistryCl
...
-    ???
-    //    new BasedOnVersionAvroSchemaDeterminer(schemaRegistryClient, preparedTopic.prepared, version, isKey = false)
+    new BasedOnVersionParsedSchemaDeterminer(schemaRegistryClient, preparedTopic.prepared, version, isKey = false)
   }
 
   //TODO: add schema versioning for key schemas
   private def prepareUniversalKeySchemaDeterminer(preparedTopic: PreparedKafkaTopic): ParsedSchemaDeterminer = {
-    ???
-    //    new BasedOnVersionAvroSchemaDeterminer(schemaRegistryClient, preparedTopic.prepared, LatestSchemaVersion, isKey = true)
+    new BasedOnVersionParsedSchemaDeterminer(schemaRegistryClient, preparedTopic.prepared, LatestSchemaVersion, isKey = true)
```
[the remark this answers is not in the document]
Comment (gadomsky) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/source/UniversalKafkaSourceFactory.scala]: No, cleaned up

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalMessageReader.scala (hunk 3 above)
Comment (gadomsky) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalMessageReader.scala:21]: Agree, but I'll do such refactor in one of the next PR's

Comment (mproch) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/UniversalMessageReader.scala:21]: :+1:

**Review by mproch — APPROVED** [review]

**Pull request merged**